components/net/hsts.rs

```diff
@@ -139,15 +142,160 @@ impl HstsList {
     }
 
     /// Step 10 of https://fetch.spec.whatwg.org/#concept-main-fetch.
-    pub fn switch_known_hsts_host_domain_url_to_https(&self, url: &mut ServoUrl) {
-        if url.scheme() != "http" {
+    pub fn apply_hsts_rules(&self, url: &mut ServoUrl) {
+        if url.scheme() != "http" && url.scheme() != "ws" {
             return;
         }
-        if url
-            .domain()
-            .map_or(false, |domain| self.is_host_secure(domain))
+
+        let mut upgrade_scheme = false;
+
+        if pref!(network.enforce_tls.enabled) {
+            upgrade_scheme = true;
+            if !pref!(network.enforce_tls.localhost) {
+                let is_localhost = match url.host() {
+                    Some(Host::Domain(domain)) => {
+                        domain.ends_with(".localhost") || domain == "localhost"
+                    },
+                    Some(Host::Ipv4(ipv4)) => ipv4.is_loopback(),
+                    Some(Host::Ipv6(ipv6)) => ipv6.is_loopback(),
+                    _ => false,
+                };
+                upgrade_scheme = !is_localhost;
+            }
+            if !pref!(network.enforce_tls.onion) &&
+                url.domain()
+                    .map_or(false, |domain| domain.ends_with(".onion"))
+            {
+                upgrade_scheme = false;
+            }
+        }
+
+        if !upgrade_scheme &&
+            url.domain()
+                .map_or(false, |domain| self.is_host_secure(domain))
         {
-            url.as_mut_url().set_scheme("https").unwrap();
+            upgrade_scheme = true;
+        }
+
+        if upgrade_scheme {
+            let upgraded_scheme = match url.scheme() {
+                "ws" => "wss",
+                _ => "https",
+            };
+            url.as_mut_url().set_scheme(upgraded_scheme).unwrap();
         }
     }
+
+    pub fn update_hsts_list_from_response(&mut self, url: &ServoUrl, headers: &HeaderMap) {
+        if url.scheme() != "https" && url.scheme() != "wss" {
+            return;
+        }
+
+        if let Some(header) = headers.typed_get::<StrictTransportSecurity>() {
+            if let Some(host) = url.domain() {
+                let include_subdomains = if header.include_subdomains {
+                    IncludeSubdomains::Included
+                } else {
+                    IncludeSubdomains::NotIncluded
+                };
+
+                if let Some(entry) =
+                    HstsEntry::new(host.to_owned(), include_subdomains, Some(header.max_age))
+                {
+                    info!("adding host {} to the strict transport security list", host);
+                    info!("- max-age {}", header.max_age);
+                    if header.include_subdomains {
+                        info!("- includeSubdomains");
+                    }
+
+                    self.push(entry);
+                }
+            }
+        }
+    }
+}
+
+// TODO: Remove this when the headers crate exposes STS struct fields
```

Comment:
File an issue for this in the headers repo, and link to it here?

Reply:
https://github.com/hyperium/headers/issues/61

Reply:
Thanks!

Reply:
Can you add a link to the hyper issue?

Reply:
Done, and I will create a PR to add getter methods.